File: python/day01_trebuchet.py
# Day 1: Trebuchet?1
# https://adventofcode.com/2023/day/1

from typing import Tuple, List
from util.input_util import read_input_file

def parse_lines() -> List[str]:
    lines = read_input_file(1)
    return lines

def is_digit(c:str):
    return '0' <= c <= '9'

def first_last_digit(line:str) -> Tuple[int, int]:
    first = None
    last = None
    for c in line:
        if is_digit(c):
            if not first:
                first = int(c)
            last = int(c)
    
    return first, last


def solution1():
    lines = parse_lines()

    s = 0
    for l in lines:
        first, last = first_last_digit(l)
        s += first * 10 + last
    return s
        

word2int = {'zero': 0,
            'one': 1,
            'two': 2,
            'three': 3,
            'four': 4,
            'five': 5,
            'six': 6,
            'seven': 7,
            'eight': 8,
            'nine':9}


def first_last_digit_words(line:str) -> Tuple[int, int]:
    first = None
    last = None
    for i, c in enumerate(line):
        if is_digit(c):
            if not first:
                first = int(c)
            last = int(c)

        for j in range(i+1, min(i+5, len(line))+1):
            ss = line[i:j]
            if ss in word2int:
                if not first:
                    first = word2int[ss]
                last = word2int[ss]

    return first, last
        

    
def solution2():
    lines = parse_lines()

    s = 0
    for l in lines:
        first, last = first_last_digit_words(l)
        s += first * 10 + last
    return s 
    
    
if __name__ == '__main__':
    print(solution1())
    
    print('--------------')
    
    print(solution2())

File: cpp/day01_trebuchet.cpp
// Day 1: Trebuchet?!
// https://adventofcode.com/2023/day/1
#include<iostream>
#include<fstream>
#include<set>
#include<cmath>
#include<string>

using namespace std;

auto read_input() {
    string filename = "../inputs/day01_input.txt";
    cout << "Input file: " << filename << endl;
    string line;
    vector<string> lines;
    ifstream input(filename);
    while (getline(input, line)) {
        lines.push_back(line);
    }
    input.close();
    return lines;
}

// combine first digit and last digit to form a single two digit number

pair<int, int> first_and_last_digit(string s) {
    int first = -1;
    int last = -1;
    for (char c : s) {
        if (c >= '0' and c <= '9') {
            if (first == -1) {
                first = c - '0';
            }
            last = c - '0';
        }
    }
    return make_pair(first, last);
}

int part1() {
    vector<string> lines = read_input();

    int sum = 0;
    for (string s : lines) {
        pair<int, int> digits = first_and_last_digit(s);
        sum += digits.first * 10 + digits.second;
    }
    return sum;
}


unordered_map<string, int> word2int = {
    {"zero", 0},
    {"one", 1},
    {"two", 2},
    {"three", 3},
    {"four", 4},
    {"five", 5},
    {"six", 6},
    {"seven", 7},
    {"eight", 8},
    {"nine", 9}
};

pair<int, int> first_and_last_digit_words(string s) {
    int first = -1;
    int last = -1;
    string ss;
    char c;
    for (int i=0; i<s.length(); i++) {
        // check first if digit
        c = s[i];
        if (c >= '0' and c <= '9') {
            if (first == -1) {
                first = c - '0';
            }
            last = c - '0';
        }
        // else see if is a start of a word
        // check for substring with next 5 chars
        for (int j=1; j<6; j++) {
            ss = s.substr(i, j);
            if (word2int.count(ss)) {
                // cout << ss << endl;
                if (first == -1) {
                    first = word2int[ss];
                }
                last = word2int[ss];
            }
        }
    }
    return make_pair(first, last);
}


int part2() {

    vector<string> lines = read_input();

    int sum = 0;
    for (string s : lines) {
        pair<int, int> digits = first_and_last_digit_words(s);
        sum += digits.first * 10 + digits.second;
    }
    return sum;
}


int main() {
    auto p1 = part1();
    cout << "Part 1: " << p1 << endl;

    cout << "--------------" << endl;

    auto p2 = part2();
    cout << "Part 2: " << p2 << endl;
}